# Importing libraries

In [1]:
import pandas as pd
import numpy as np
import os

In [2]:
path = r'C:\Users\Dan\Desktop\Career Foundry\Instacart Basket Analysis'

In [3]:
df = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'customers.csv'), index_col = False)

In [4]:
#checking dataframe
df.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [5]:
df.shape

(206209, 10)

In [6]:
# rename column
df.rename(columns = {'Surnam' : 'Surname'}, inplace = True)

In [7]:
df.rename(columns = {'n_dependants' : 'no._of_dependants'}, inplace = True)

In [8]:
df.rename(columns = {'fam_status' : 'marital_status'}, inplace = True)

In [9]:
df.head()

user_id First Name   Surname  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   no._of_dependants marital_status  income  
0                  3        married  165665  
1                  0         single   59285  
2                  2        married   99568  
3                  0         single   42049  
4                  1        married   40374

In [10]:
#checking data types
df.dtypes

user_id               int64
First Name           object
Surname              object
Gender               object
STATE                object
Age                   int64
date_joined          object
no._of_dependants     int64
marital_status       object
income                int64
dtype: object

In [11]:
# changing user_id to string
df['user_id'] = df['user_id'].astype('str')

In [12]:
#check
df.dtypes

user_id              object
First Name           object
Surname              object
Gender               object
STATE                object
Age                   int64
date_joined          object
no._of_dependants     int64
marital_status       object
income                int64
dtype: object

In [13]:
# descriptive stats
df.describe()

Age  no._of_dependants         income
count  206209.000000      206209.000000  206209.000000
mean       49.501646           1.499823   94632.852548
std        18.480962           1.118433   42473.786988
min        18.000000           0.000000   25903.000000
25%        33.000000           0.000000   59874.000000
50%        49.000000           1.000000   93547.000000
75%        66.000000           3.000000  124244.000000
max        81.000000           3.000000  593901.000000

In [14]:
# checking for missing values
df.isnull().sum()

user_id                  0
First Name           11259
Surname                  0
Gender                   0
STATE                    0
Age                      0
date_joined              0
no._of_dependants        0
marital_status           0
income                   0
dtype: int64

this shows there are 11,259 cases of no entry under first name 

In [15]:
# changing missing values to NaN
df_nan = df[df['First Name'].isnull() == True]

In [16]:
df_nan

user_id First Name   Surname  Gender         STATE  Age date_joined  \
53       76659        NaN   Gilbert    Male      Colorado   26    1/1/2017   
73       13738        NaN     Frost  Female     Louisiana   39    1/1/2017   
82       89996        NaN    Dawson  Female        Oregon   52    1/1/2017   
99       96166        NaN   Oconnor    Male      Oklahoma   51    1/1/2017   
105      29778        NaN    Dawson  Female          Utah   63    1/1/2017   
...        ...        ...       ...     ...           ...  ...         ...   
206038  121317        NaN    Melton    Male  Pennsylvania   28   3/31/2020   
206044  200799        NaN  Copeland  Female        Hawaii   52    4/1/2020   
206090  167394        NaN     Frost  Female        Hawaii   61    4/1/2020   
206162  187532        NaN     Floyd  Female    California   39    4/1/2020   
206171  116898        NaN   Delgado  Female      Colorado   23    4/1/2020   

        no._of_dependants marital_status  income  
53                      2        married   41709  
73                      0         single   82518  
82                      3        married  117099  
99                      1        married  155673  
105                     3        married  151819  
...                   ...            ...     ...  
206038                  3        married   87783  
206044                  2        married  108488  
206090                  1        married   45275  
206162                  0         single   56325  
206171                  2        married   59222  

[11259 rows x 10 columns]

In [17]:
df.isnull().sum()

user_id                  0
First Name           11259
Surname                  0
Gender                   0
STATE                    0
Age                      0
date_joined              0
no._of_dependants        0
marital_status           0
income                   0
dtype: int64

In [18]:
#Check for data consistentcy - mixed types
for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

First Name


In [19]:
#Changing 'First Name' to a string
df['First Name'] = df['First Name'].astype('str')

In [20]:
df.dtypes

user_id              object
First Name           object
Surname              object
Gender               object
STATE                object
Age                   int64
date_joined          object
no._of_dependants     int64
marital_status       object
income                int64
dtype: object

In [21]:
#checking again for missing values
df.isnull().sum()

user_id              0
First Name           0
Surname              0
Gender               0
STATE                0
Age                  0
date_joined          0
no._of_dependants    0
marital_status       0
income               0
dtype: int64

no missing values now show up under First Name as the nans are now strings

In [22]:
#Checking for duplicates
df_dups = df[df.duplicated()]

In [23]:
df_dups.shape

(0, 10)

In [24]:
df_dups

Empty DataFrame
Columns: [user_id, First Name, Surname, Gender, STATE, Age, date_joined, no._of_dependants, marital_status, income]
Index: []

dataframe is empty, therefore no duplicates

# importing dataset


In [25]:
ord_pro_mer = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged_small_sample.pkl'))

In [26]:
#checking date types
ord_pro_mer.dtypes

order_id                    int32
user_id                     int32
order_number                 int8
orders_day_of_week           int8
order_hour_of_day            int8
days_since_last_order     float16
product_id                  int32
add_to_cart_order           int32
reordered                    int8
product_name               object
aisle_id                     int8
department_id                int8
prices                    float16
_merge                   category
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                    int8
loyalty_flag               object
average_price             float32
spending_flag              object
median_last_orders        float16
order_frequency_flag       object
dtype: object

In [27]:
# both datasets contains user_id columns so will change user_id date type to string to match that in customers dataset
ord_pro_mer['user_id'] = ord_pro_mer['user_id'].astype('str')

In [28]:
ord_pro_mer.dtypes

order_id                    int32
user_id                    object
order_number                 int8
orders_day_of_week           int8
order_hour_of_day            int8
days_since_last_order     float16
product_id                  int32
add_to_cart_order           int32
reordered                    int8
product_name               object
aisle_id                     int8
department_id                int8
prices                    float16
_merge                   category
price_range_loc            object
busiest_day                object
busiest_days               object
busiest_period_of_day      object
max_order                    int8
loyalty_flag               object
average_price             float32
spending_flag              object
median_last_orders        float16
order_frequency_flag       object
dtype: object

In [29]:
#combining the data sets on the 'user_id' column df_combined = ord_pro_mer.merge(df, on = 'user_id') crashed my computer many times after a wait of hours. 

In [30]:
df_merged = ord_pro_mer.merge(df, on = 'user_id')

In [31]:
#after many attempts at merging the dataframes in different way I have finally managed to merge a 30% sample with the customers dataframe and will continue with that 

In [32]:
df_merged . shape

(9721098, 33)

In [33]:
df_merged . head()

order_id user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329       1             1                   2                  8   
1    473747       1             3                   3                 12   
2   2254736       1             4                   4                  7   
3    550135       1             7                   1                  9   
4   2539329       1             1                   2                  8   

   days_since_last_order  product_id  add_to_cart_order  reordered  \
0                    NaN         196                  1          0   
1                   21.0         196                  1          1   
2                   29.0         196                  1          1   
3                   20.0         196                  1          1   
4                    NaN       12427                  3          0   

          product_name  ...   order_frequency_flag  First Name  Surname  \
0                 Soda  ...  Non-frequent customer       Linda   Nguyen   
1                 Soda  ...  Non-frequent customer       Linda   Nguyen   
2                 Soda  ...  Non-frequent customer       Linda   Nguyen   
3                 Soda  ...  Non-frequent customer       Linda   Nguyen   
4  Original Beef Jerky  ...  Non-frequent customer       Linda   Nguyen   

   Gender    STATE Age date_joined no._of_dependants  marital_status income  
0  Female  Alabama  31   2/17/2019                 3         married  40423  
1  Female  Alabama  31   2/17/2019                 3         married  40423  
2  Female  Alabama  31   2/17/2019                 3         married  40423  
3  Female  Alabama  31   2/17/2019                 3         married  40423  
4  Female  Alabama  31   2/17/2019                 3         married  40423  

[5 rows x 33 columns]

In [34]:
# exporting dataset
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'ord_pro_cus_merged.pkl'))